### PR TITLE
unify code formatting

### PR DESCRIPTION
**Pull request by Blef666**
This is a proposal for a unified code formatting. The configuration is stored in `.editorconfig` file so anyone can change it later. If you've got any subjective, estetical remarks, be welcome to share them, I'll consider involving them.
Currently there is no possibility of creating solution consisting of both automatic code formatting in IDE and CI pipeline validation. I limited it to only IDE formatting.

### CODE REVIEW

**Comment by adamw** [comment]
No remark, any formatting is good as long as it's consistent.

Just two questions:
1. does the editorconfig file has to be so big? is it somehow automatically generated, or where does it come from?
2. does the formatting reflect the defaults in IntelliJ, or is it somehow already customized?

File: flows/src/test/java/com/softwaremill/jox/flows/FlowFlattenTest.java [hunk 1]
```diff
@@ -99,14 +99,14 @@ void shouldHandleSubsequentFlattenCalls() throws Exception {
 
         // when & then
         var result = flow.flattenPar(10)
-                .runToList().stream().flatMap(f -> {
+                         .runToList().stream().flatMap(f -> {
```
Comment (emil-bar) [inline flows/src/test/java/com/softwaremill/jox/flows/FlowFlattenTest.java:102]: It looks at least werid that this line and 109 have got additional indent, but lines between didn't (check yourself in IJ)

File: flows/src/test/java/com/softwaremill/jox/flows/FlowGroupedTest.java [hunk 2]
```diff
@@ -162,9 +162,9 @@ void groupedWithin_shouldGroupFirstBatchOfElementsDueToTimeoutAndSecondBatchDueT
 
             // when
             var elementsWithEmittedTimeOffset = Flows.fromSource(c)
-                    .groupedWithin(3, Duration.ofMillis(100))
-                    .map(s -> Map.entry(s, Duration.ofNanos(System.nanoTime() - start)))
-                    .runToList();
+                                                     .groupedWithin(3, Duration.ofMillis(100))
```
Comment (emil-bar) [inline flows/src/test/java/com/softwaremill/jox/flows/FlowGroupedTest.java:165]: So I believe previous version of indent is better (here and in other places as well), as it leaves you more spaces to write the line

File: .editorconfig [hunk 3]
```diff
@@ -0,0 +1,1232 @@
+[*]
+charset = utf-8
+end_of_line = lf
+indent_size = 4
+indent_style = space
+insert_final_newline = false
+max_line_length = 120
```
[the remark this answers is not in the document]
Comment (Blef666) [reply .editorconfig:7]: 120 is a default value in idea, but maybe it's better to increase it to 140. I'll do that

File: README.md [hunk 4]
```diff
@@ -120,8 +120,8 @@ class Demo2 {
 }
 ```
 
-There is also a possibility to pass `Channel`'s buffer size via `ScopedValue`. 
-The channel must be created via `Channel.withScopedBufferSize()` to get the value. 
+There is also a possibility to pass `Channel`'s buffer size via `ScopedValue`.
```
[the remark this answers is not in the document]
Comment (Blef666) [reply README.md:123]: Well, it does look better with line breaks so I'd leave it. Especially that in `*.md` files it's common to have a really long lines.

File: channels/src/test/java/com/softwaremill/jox/ChannelInterruptionTest.java [hunk 5]
```diff
@@ -7,8 +7,8 @@
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Semaphore;
 
-import static java.util.concurrent.TimeUnit.SECONDS;
 import static com.softwaremill.jox.TestUtil.*;
+import static java.util.concurrent.TimeUnit.SECONDS;
```
[the remark this answers is not in the document]
Comment (Blef666) [reply channels/src/test/java/com/softwaremill/jox/ChannelInterruptionTest.java:11]: I left the default imports order from Intellij:
1. custom
2. javax
3. java
4. statics

File: channels/src/test/java/com/softwaremill/jox/ChannelInterruptionTest.java [hunk 6]
```diff
@@ -7,8 +7,8 @@
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Semaphore;
 
-import static java.util.concurrent.TimeUnit.SECONDS;
 import static com.softwaremill.jox.TestUtil.*;
```
[the remark this answers is not in the document]
Comment (Blef666) [reply channels/src/test/java/com/softwaremill/jox/ChannelInterruptionTest.java:10]: Wildcards are used when there are at least 5  classes from a package or 3 static names. It's the default Intellij configuration.

File: channels/src/test/java/com/softwaremill/jox/ChannelInterruptionTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamw) [reply channels/src/test/java/com/softwaremill/jox/ChannelInterruptionTest.java:11]: I'd generally leave IJ defaults ;)

File: flows/src/test/java/com/softwaremill/jox/flows/FlowGroupedTest.java (hunk 2 above)
Comment (Blef666) [reply flows/src/test/java/com/softwaremill/jox/flows/FlowGroupedTest.java:165]: It's because of the `Align when multiline` for chained method calls. It's easier to determine which nesting level in method calls we are in, when particular levels are in the same column. It's for the nested method calls.
Previous version was better for long lines, but this one is for shorter lines. There are way more shorter lines using chained method calls in the repo then longer so I'd leave it.

File: flows/src/test/java/com/softwaremill/jox/flows/FlowFlattenTest.java (hunk 1 above)
Comment (Blef666) [reply flows/src/test/java/com/softwaremill/jox/flows/FlowFlattenTest.java:102]: The change is because `Align when multiline` for chained method calls is turned on. I can't find an option for also adding indents for the blocks inside of the method call including lambda functions. It looks like the lines 103-107 might be a non covered edge case.

Comment (adamw) [reply flows/src/test/java/com/softwaremill/jox/flows/FlowFlattenTest.java:102]: That's exactly why we should simply stick to the default IJ config. No customisations - just the defaults. It cuts all formatting discussions at their roots ;)

**Comment by Blef666** [comment]
@adamw 
1. `.editorconfig` is that big because it's exported form Intellij. Anyone who makes code style changes in the IDE and exports them will get the same size so cutting it only to changes made wouldn't be feasible.
2. Everything is a default for Intellij except for 2 custom configs: 
- alignment in multilne chain of method calls
- keeping simple blocks `{}` in one line
- and maybe I'll increase line length from 120 to 140

@emil-bar As I explained, cutting the `.editorconfig` file only to relevant file types wouldn't make any difference because it's default configuration for Intellij. However, it would introduce additional work.

**Comment by adamw** [comment]
Ok, well you know my opinion, I'd just keep IJ defaults ;)

File: .editorconfig (hunk 3 above)
[the remark this answers is not in the document]
Comment (Blef666) [reply .editorconfig:7]: I left 120.

File: flows/src/test/java/com/softwaremill/jox/flows/FlowFlattenTest.java (hunk 1 above)
Comment (Blef666) [reply flows/src/test/java/com/softwaremill/jox/flows/FlowFlattenTest.java]: I switched back to the default configuration.

**Comment by Blef666** [comment]
In the end I switched back to the default Intellij configuration except for the single line simple blocks `{}` I think no one argues it looks better.
